connector/routingconnector/logs.go

```diff
@@ -0,0 +1,145 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routingconnector // import "github.com/open-telemetry/opentelemetry-collector-contrib/connector/routingconnector"
+
+import (
+	"context"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/connector"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/connector/routingconnector/internal/common"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottllog"
+)
+
+type logsConnector struct {
+	component.StartFunc
+	component.ShutdownFunc
+
+	logger *zap.Logger
+	config *Config
+	router *router[consumer.Logs, ottllog.TransformContext]
+}
+
+type logsGroup struct {
+	consumer consumer.Logs
+	logs     plog.Logs
+}
+
+func newLogsConnector(
+	set connector.CreateSettings,
+	config component.Config,
+	logs consumer.Logs,
+) (*logsConnector, error) {
+	cfg := config.(*Config)
+
+	logParser, _ := ottllog.NewParser(
+		common.Functions[ottllog.TransformContext](),
+		set.TelemetrySettings,
+	)
+
+	lr, ok := logs.(connector.LogsRouter)
+	if !ok {
+		return nil, errTooFewPipelines
+	}
+
+	r, err := newRouter(
+		cfg.Table,
+		cfg.DefaultPipelines,
+		lr.Consumer,
+		set.TelemetrySettings,
+		logParser)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &logsConnector{
+		logger: set.TelemetrySettings.Logger,
+		config: cfg,
+		router: r,
+	}, nil
+}
+
+func (c *logsConnector) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
+}
+
+func (c *logsConnector) ConsumeLogs(ctx context.Context, ld plog.Logs) error {
+	// routingEntry is used to group plog.ResourceLogs that are routed to
+	// the same set of exporters.
+	// This way we're not ending up with all the logs split up which would cause
+	// higher CPU usage.
+	groups := map[string]logsGroup{}
+	var errs error
+
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		rlogs := ld.ResourceLogs().At(i)
+		ltx := ottllog.NewTransformContext(
+			plog.LogRecord{},
+			pcommon.InstrumentationScope{},
+			rlogs.Resource(),
+		)
```

Comment:
Should we be using `ottlresource.NewTransformContext`?

Reply:
I made this change in c7e8a2bf74261f8731bda4d17eaf99602dbe4a37. Instead of using `ottllog.TransformContext`, `ottlspan.TransformContext`, etc we're now using `ottlresource.TransformContext` everywhere. This allowed me to remove the second type parameter from the router. It also means our statements can refer to `attributes` instead of `resource.attributes`. If you don't mind, take a look at the commit and let me know if that's what you had in mind. It simplifies the code significantly, so long as we're not losing anything. The choice to use the signal specific `TransformContext`s was inherited from the routing processor.

Reply:
cc: @jpkrohling @kovrus to double check this change

Reply:
Looks right to me. We'll end up needing the signal specific contexts for other conditions but for routing based on resources, I think we can and should keep it simple.